crates/nu-parser/src/lex/tokens.rs

```diff
@@ -0,0 +1,212 @@
+use derive_new::new;
+use itertools::Itertools;
+use std::fmt;
+
+use nu_source::{HasSpan, Span, Spanned, SpannedItem};
+
+use super::token_group::TokenBuilder;
+
+#[derive(Debug, Clone, PartialEq, is_enum_variant)]
+pub enum TokenContents {
+    /// A baseline token is an atomic chunk of source code. This means that the
+    /// token contains the entirety of string literals, as well as the entirety
+    /// of sections delimited by paired delimiters.
+    ///
+    /// For example, if the token begins with `{`, the baseline token continues
+    /// until the closing `}` (after taking comments and string literals into
+    /// consideration).
+    Baseline(String),
+    Comment(LiteComment),
+    Pipe,
+    Semicolon,
+    EOL,
+}
+
+impl fmt::Display for TokenContents {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            TokenContents::Baseline(base) => write!(f, "{}", base),
+            TokenContents::Comment(comm) => write!(f, "{}", comm),
+            TokenContents::Pipe => write!(f, "|"),
+            TokenContents::Semicolon => write!(f, ";"),
+            TokenContents::EOL => write!(f, "\\n"),
+        }
+    }
+}
+
+pub type CommandBuilder = TokenBuilder<Spanned<String>>;
+pub type CommentsBuilder = TokenBuilder<LiteComment>;
+pub type PipelineBuilder = TokenBuilder<LiteCommand>;
+pub type GroupBuilder = TokenBuilder<PipelineBuilder>;
+
+/// A LiteComment is a line comment. It begins with `#` and continues until (but not including) the
+/// next newline.
+///
+/// It remembers any leading whitespace, which is used in later processing steps to strip off
+/// leading whitespace for an entire comment block when it is associated with a definition.
+#[derive(Debug, PartialEq, Clone)]
+pub struct LiteComment {
+    leading_ws: Option<Spanned<String>>,
+    rest: Spanned<String>,
+}
+
+impl LiteComment {
+    pub fn new(string: impl Into<Spanned<String>>) -> LiteComment {
+        LiteComment {
+            leading_ws: None,
+            rest: string.into(),
+        }
+    }
+
+    pub fn new_with_ws(
+        ws: impl Into<Spanned<String>>,
+        comment: impl Into<Spanned<String>>,
+    ) -> LiteComment {
+        LiteComment {
+            leading_ws: Some(ws.into()),
+            rest: comment.into(),
+        }
+    }
+
+    pub fn dedent(&self, excluded_spaces: usize) -> LiteComment {
```

Comment:
As a non native english speaker, it is hard for me to figure out what "dedent" means here. Maybe a better name would be: "trim_max"?

Reply:
or unindent